crypto/ocsp/ocsp_http.c

```diff
@@ -22,11 +26,335 @@
+            return 0;
+        }
+
+        // First line.
+        if (rctx->state == OHS_FIRSTLINE) {
+            if (parse_http_line((char *)rctx->iobuf)) {
+                rctx->state = OHS_HEADERS;
+                goto next_line;
+            } else {
+                rctx->state = OHS_ERROR;
+                return 0;
+            }
+        } else {
+            // Look for blank line: end of headers.
+            for (p = rctx->iobuf; *p; p++) {
+                if ((*p != '\r') && (*p != '\n')) {
+                  break;
+                }
+            }
+            if (*p != '\0') {
+              goto next_line;
+            }
+
+            rctx->state = OHS_ASN1_HEADER;
+        }
+
+        OPENSSL_FALLTHROUGH;
+    case OHS_ASN1_HEADER:
+        // Now reading ASN1 header: can read at least 2 bytes which is enough
+        // for ASN1 SEQUENCE header and either length field or at least the
+        // length of the length field.
+        n = BIO_get_mem_data(rctx->mem, &p);
+        if (n < 2) {
+          goto next_io;
+        }
+
+        // Check it is an ASN1 SEQUENCE.
+        if (*p++ != (V_ASN1_SEQUENCE | V_ASN1_CONSTRUCTED)) {
+            rctx->state = OHS_ERROR;
+            return 0;
+        }
+
+        // Check out length field.
+        if ((*p & 0x80) != 0) {
+            // If MSB set on initial length octet we can now always read 6
+            // octets: make sure we have them.
+            if (n < 6) {
+              goto next_io;
+            }
+            n = *p & 0x7F;
+            // Not NDEF or excessive length.
+            if (!n || (n > 4)) {
+                rctx->state = OHS_ERROR;
+                return 0;
+            }
+            p++;
+            rctx->asn1_len = 0;
+            for (i = 0; i < n; i++) {
+                rctx->asn1_len <<= 8;
+                rctx->asn1_len |= *p++;
+            }
+            if (rctx->asn1_len > rctx->max_resp_len) {
+                rctx->state = OHS_ERROR;
+                return 0;
+            }
+            rctx->asn1_len += n + 2;
+        } else {
+          rctx->asn1_len = *p + 2;
+        }
+        rctx->state = OHS_ASN1_CONTENT;
+
+        OPENSSL_FALLTHROUGH;
+    case OHS_ASN1_CONTENT:
+        n = BIO_get_mem_data(rctx->mem, NULL);
+        if (n < (int)rctx->asn1_len) {
+          goto next_io;
+        }
+        rctx->state = OHS_DONE;
+        return 1;
+
+    case OHS_DONE:
+        return 1;
+    }
+
+    return 0;
+}
+
+int OCSP_sendreq_nbio(OCSP_RESPONSE **presp, OCSP_REQ_CTX *rctx)
+{
+    return OCSP_REQ_CTX_nbio_d2i(rctx, (ASN1_VALUE **)presp,
+                                 ASN1_ITEM_rptr(OCSP_RESPONSE));
+}
+
+OCSP_RESPONSE *OCSP_sendreq_bio(BIO *b, const char *path, OCSP_REQUEST *req)
+{
+    OCSP_RESPONSE *resp = NULL;
+    OCSP_REQ_CTX *ctx;
+    int rv;
+
+    ctx = OCSP_sendreq_new(b, path, req, -1);
+    if (ctx == NULL) {
+      return NULL;
+    }
+
+    // This waits indefinitely on a response.
```

Comment:
I assume this comment is 100% not accurate, I assume a socket timeout or some other error could cause a break-out this loop?

Reply:
This waits indefinitely if `BIO_should_retry` is true, clarified the comment. Also added a documentation comment clarifying this API is only left for compatibility and people should avoid using it.